### PR TITLE
Add StreamFactoryDiscovery

### PR DESCRIPTION
**Pull request by mekras**
Supports Guzzle and Zend Diactoros.

### CODE REVIEW

File: src/StreamFactory/DiactorosStreamFactory.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+use Http\Message\StreamFactory;
+use Psr\Http\Message\StreamInterface;
+use Zend\Diactoros\Request;
+use Zend\Diactoros\Response;
+use Zend\Diactoros\Stream;
+
+/**
+ * Zend Diactoros StreamFactory
+ *
+ * @author GeLo <geloen.eric@gmail.com>
```
Comment (sagikazarmark) [inline src/StreamFactory/DiactorosStreamFactory.php]: Remove Eric please.

File: src/StreamFactory/DiactorosStreamFactory.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+namespace Http\Discovery\StreamFactory;
+
+use Http\Message\StreamFactory;
+use Psr\Http\Message\StreamInterface;
+use Zend\Diactoros\Request;
+use Zend\Diactoros\Response;
+use Zend\Diactoros\Stream;
+
+/**
+ * Zend Diactoros StreamFactory
```
Comment (sagikazarmark) [inline src/StreamFactory/DiactorosStreamFactory.php]: This is not necessary

File: src/StreamFactory/GuzzleStreamFactory.php [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Http\Discovery\StreamFactory;
+
+use Http\Message\StreamFactory;
+use Psr\Http\Message\StreamInterface;
+
+/**
+ * Guzzle Stream Factory
```
Comment (sagikazarmark) [inline src/StreamFactory/GuzzleStreamFactory.php]: This is not necessary

**Comment by sagikazarmark** [comment]
Nice job @mekras

File: src/StreamFactory/DiactorosStreamFactory.php (hunk 1 above)
Comment (mekras) [inline src/StreamFactory/DiactorosStreamFactory.php]: But it's code written by him. I just copied it from MessageFactory.

Comment (sagikazarmark) [inline src/StreamFactory/DiactorosStreamFactory.php]: Actually, it is written by me. Back then I added Eric to each files he had any connection with, but that class is 90% rewritten by me. Even if it is copied partly, please add only yourself.

You might also add a comment that the code is copied from message factory. Actually, the ultimate solution would be somehow share the code between the two classes. But let's think about it later.

File: src/StreamFactory/DiactorosStreamFactory.php (hunk 2 above)
Comment (mekras) [inline src/StreamFactory/DiactorosStreamFactory.php]: Not necessarily, but also not forbidden, right?

File: src/StreamFactory/GuzzleStreamFactory.php (hunk 3 above)
Comment (mekras) [inline src/StreamFactory/GuzzleStreamFactory.php]: Not necessarily, but also not forbidden, right?

Comment (sagikazarmark) [inline src/StreamFactory/GuzzleStreamFactory.php]: No, but I consider it to be noise as there is no added value in it. (It is exactly the same as the classname) I am going to remove it from other factories as well.

File: src/StreamFactory/DiactorosStreamFactory.php (hunk 1 above)
Comment (mekras) [inline src/StreamFactory/DiactorosStreamFactory.php]: OK

**Comment by joelwurtz** [comment]
I think message factory should have an injected stream factory to create stream (actually it create stream from a protected method)

**Comment by sagikazarmark** [comment]
I also thought about that. But it would complicate using the MessageFactory.

Actually, we could even instantiate the stream factory or use the StreamDiscovery in MessageFactory.

Since the MessageFactory uses the exact same code, hard coupling is probably not a problem here.

**Pull request merged**